Enabling MediaSourcePrefersDecompressionSession cause 60fps playback of 24fps video in Youtube

jyavenard opened the pull request:
#### 36dd4bd72f202a31bd837012d70b2accd4891230
<pre>
Enabling MediaSourcePrefersDecompressionSession cause 60fps playback of 24fps video in Youtube
<a href="https://bugs.webkit.org/show_bug.cgi?id=292238">https://bugs.webkit.org/show_bug.cgi?id=292238</a>
<a href="https://rdar.apple.com/149335351">rdar://149335351</a>

Reviewed by Eric Carlson and Jer Noble.

AVSampleBufferVideoRendererPowerOptimization requires two things to be fully effective:
- Enqueue ahead of currentTime multiple samples.
- regularly call [AVSBDL expectMinimumUpcomingSampleBufferPresentationTime]

We expand on 294144@main and 294057@main to comply with those two requirements by
calling expectMinimumUpcomingSampleBufferPresentationTime with the next sample time
that is about to be enqueued.

Manually verified that the WindowServer drops the screen refresh rate to as low as 24fps.

* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.h:
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
(WebCore::VideoMediaSampleRenderer::hasIncomingOutOfOrderFrame const):
(WebCore::VideoMediaSampleRenderer::minimumUpcomingSampleTime const):
(WebCore::VideoMediaSampleRenderer::decodeNextSampleIfNeeded):

Canonical link: <a href="https://commits.webkit.org/294258@main">https://commits.webkit.org/294258@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cd020f8f66df6fa072812dca1f6195c8f32f43e8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101255 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20917 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11220 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106403 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51882 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/103296 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21226 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29412 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77110 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34152 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104262 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16367 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91457 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57457 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16183 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9468 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/51230 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86066 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9543 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108759 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28383 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20888 "Found 1 new test failure: imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-default-writing-mode-rl.html (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86087 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-type-on-non-root.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28745 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87665 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85640 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21795 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30371 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8080 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/22482 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28313 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33584 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28125 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31445 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29683 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -252,6 +255,28 @@ static OSStatus sampleCallback(CMBufferRef buffer, void* refcon)
+
+    // The maximum queue depth possible for out of order frames with either H264 or HEVC is 16, limit looking ahead of 16 frames.
+    size_t forwardIndex = 0;
+    for (auto it = m_compressedSampleQueue.begin(); it != m_compressedSampleQueue.end() && forwardIndex < 16; ++forwardIndex, ++it) {
```
Nit: it would be good to define a `constexpr` instead of hardcoding 16 since the same logic is in `hasIncomingOutOfOrderFrame`.

jyavenard committed:
Enabling MediaSourcePrefersDecompressionSession cause 60fps playback of 24fps video in Youtube
https://bugs.webkit.org/show_bug.cgi?id=292238
rdar://149335351

Reviewed by Eric Carlson and Jer Noble.

AVSampleBufferVideoRendererPowerOptimization requires two things to be fully effective:
- Enqueue ahead of currentTime multiple samples.
- regularly call [AVSBDL expectMinimumUpcomingSampleBufferPresentationTime]

We expand on 294144@main and 294057@main to comply with those two requirements by
calling expectMinimumUpcomingSampleBufferPresentationTime with the next sample time
that is about to be enqueued.

Manually verified that the WindowServer drops the screen refresh rate to as low as 24fps.

* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.h:
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
(WebCore::VideoMediaSampleRenderer::hasIncomingOutOfOrderFrame const):
(WebCore::VideoMediaSampleRenderer::minimumUpcomingSampleTime const):
(WebCore::VideoMediaSampleRenderer::decodeNextSampleIfNeeded):

Canonical link: https://commits.webkit.org/294258@main
